src/Tools/dotnet-openapi/src/Commands/BaseCommand.cs

```diff
@@ -0,0 +1,290 @@
+        }
+
+        internal void AddServiceReference(
+            string tagName,
+            FileInfo projectFile,
+            string sourceFile,
+            string sourceUrl = null)
+        {
+            var project = LoadProject(projectFile);
+            var items = project.GetItems(tagName);
+            var item = items.SingleOrDefault(i => string.Equals(i.EvaluatedInclude, sourceFile));
+
+            if (sourceUrl != null)
+            {
+                var urlMatch = items.SingleOrDefault(i => string.Equals(i.GetMetadataValue(SourceUrlAttrName), sourceUrl));
+                if (urlMatch != null)
+                {
+                    Out.Write($"A reference to '{sourceUrl}' already exists in '{project.FullPath}'.");
+                    return;
+                }
+            }
+
+            if (item == null)
+            {
+                var metadata = new Dictionary<string, string>();
+
+                if (!string.IsNullOrEmpty(sourceUrl))
+                {
+                    metadata[SourceUrlAttrName] = sourceUrl;
+                }
+
+                project.AddElementWithAttributes(tagName, sourceFile, metadata);
+            }
+            else
+            {
+                Out.Write($"A reference to '{sourceFile}' already exists in '{project.FullPath}'.");
+            }
+        }
+
+        internal async Task DownloadToFileAsync(string url, string destinationPath, bool overwrite)
+        {
+            Application application;
+            if(Parent is Application)
+            {
+                application = (Application)Parent;
+            }
+            else
+            {
+                application = (Application)Parent.Parent;
+            }
+
+            var content = await application.DownloadProvider(url);
+            await WriteToFile(content, destinationPath, overwrite);
+        }
+
+        internal void EnsurePackagesInProject(FileInfo projectFile, CodeGenerator codeGenerator)
+        {
+            var packages = GetServicePackages(codeGenerator);
+            foreach (var (packageId, version) in packages)
+            {
+                var args = new string[] {
+                    "add",
+                    "package",
+                    packageId,
+                    "--version",
+                    version,
+                    "--no-restore"
+                };
+
+                var startInfo = new ProcessStartInfo
+                {
+                    FileName = DotNetMuxer.MuxerPath,
+                    Arguments = string.Join(" ", args),
+                    WorkingDirectory = projectFile.Directory.FullName,
+                    RedirectStandardError = true,
+                    RedirectStandardOutput = true,
+                };
+
+                var process = Process.Start(startInfo);
+                process.WaitForExit(20 * 1000);
+
+                if (process.ExitCode != 0)
```

Comment:
Hmm I think you should check the bool returned for WaitForExit instead of just the ExitCode in case the process didn't run to completion. In that case ExitCode will throw InvalidOperationException I believe. Also 20 seconds is a long time, since this isn't doing a restore. I would time it out earlier.

Reply:
If this actually waiting until the openapi spec has been generated then it might make sense to even increase it as it might take quite some time if there is a big API surface (ie lots of types have to be reflected, xml docs loaded etc)

Reply:
To my knowledge the spec is not generated when you add the packages we add, so I think @JunTaoLuo is right that we should reduce the timeout. I made it 5 seconds, which is still probably more than is needed, but not by an insane amount if you account for slow machines.